### ISSUE TITLE
ghcWithHoogle broken

### ISSUE THREAD

**Issue by gilligan** [NixOS/nixpkgs#17540]
## Issue description
`ghcWithHoogle` is broken

### Steps to reproduce
```nix
# config.nix
{ pkgs }:
{
  packageOverrides = pkgs : rec {
   ghcEnv = pkgs.haskellPackages.ghcWithHoogle
     (haskellPackages: with haskellPackages; [
       cabal-install
  ]);
}
```

```
% nix-env -iA nixpkgs.ghcEnv
installing ‘ghc-8.0.1’
these derivations will be built:
  /nix/store/vjn2v8wyys96q79kiwqpnn3kfi8rbl99-hoogle-local-0.1.drv
  /nix/store/mr802xsik73jizc4nbb9357985b8jkzz-ghc-8.0.1.drv
building path(s) ‘/nix/store/qwfycvpw1h31fv7yl48kwvpqsw0fdzxz-hoogle-local-0.1’
building
importing builtin packages
Error, database does not exist (run 'hoogle generate' first)
    Filename: /nix/store/amyx4gmgdj5q8zvrnidkdlnag4hs1za4-ghc-8.0.1-doc/share/doc/ghc/html/libraries/Cabal-1.24.0.0
builder for ‘/nix/store/vjn2v8wyys96q79kiwqpnn3kfi8rbl99-hoogle-local-0.1.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/mr802xsik73jizc4nbb9357985b8jkzz-ghc-8.0.1.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/mr802xsik73jizc4nbb9357985b8jkzz-ghc-8.0.1.drv’ failed
```

## Technical details

* System: OSX, Linux
* Nix version: 1.11.2
* Nixpkgs version: 16.09pre-git (382303310754e85a61eecbee4e3470abcce441ec)

**Comment by garbas** [comment]
also happens on linux

**Comment by peti** [comment]
The hoogle update to version 5 may have caused this.

**Comment by jwiegley** [comment]
Yes, it was the update to version 5.  I'm running into this as well now.  I'll look into it today.

**Comment by jwiegley** [comment]
In that case, maybe the `hoogle-local` machinery could defer to use hoogle 4 until then?

**Comment by mdorman** [comment]
That certainly seems sensible to me.

**Comment by Profpatsch** [comment]
I created a quick PR that does that.

**Comment by peti** [comment]
Please bind directly to hoogle_4_2_43 without creating an intermediate hoogle4 attribute.

**Comment by Profpatsch** [comment]
@peti Right, done.

**Comment by peti** [comment]
Fixed by https://github.com/NixOS/nixpkgs/pull/18263.

**Comment by Profpatsch** [comment]
@peti fyi I’ve been looking into hoogle5 on the Saturday of MuniHac, but it’s not yet in a workable state. The version 5 apparently doesn’t support merging of the binary databases anymore, so that part of the toolchain will have to be adjusted.

I can paste my changes so far somewhere if someone wants to look into it.

**Comment by peti** [comment]
I believe @mdorman has experimented with Hoogle 5 support before, too. Maybe he can help?

**Comment by mdorman** [comment]
@peti is correct, I collaborated with @ndmitchell on some patches to Hoogle to allow us to get this working again; I have a branch at https://github.com/mdorman/nixpkgs/tree/hoogle-fixup, though it includes some `configuration-common.nix` business to actually patch hoogle; once Neil does a new release, those are no longer necessary and I'd be comfortable merging this.

**Comment by ndmitchell** [comment]
If I made a new release tonight would that help?

**Comment by Profpatsch** [comment]
> I have a branch at https://github.com/mdorman/nixpkgs/tree/hoogle-fixup, though it includes some configuration-common.nix business to actually patch hoogle;

Nice, does that work around having to combine the database binaries afterwards?

**Comment by mdorman** [comment]
@Profpatsch Yeah, it sidesteps the whole question by having hoogle simply read everything at once. :)

@ndmitchell If you think it's ready, that would be great, but I don't think there's a reason to rush one out if you want more soak time.

**Comment by ndmitchell** [comment]
@mdorman - I've already used the patches you supplied which turned out to be required at `$WORK`, so a release benefits me too :). I've just released 5.0.4 with all your changes in it.

**Comment by mdorman** [comment]
@ndmitchell Wonderful. Then as soon as we update our master package list, I can put my change up for review.
